# Pandas

> Source: https://www.w3schools.com/python/pandas

## Resources

* CSV Data = https://www.w3schools.com/python/pandas/data.csv
* JSON Data = https://www.w3schools.com/python/pandas/data.json

## Introduction

https://www.w3schools.com/python/pandas/pandas_intro.asp

## Getting Started

https://www.w3schools.com/python/pandas/pandas_getting_started.asp

In [49]:
import pandas as pd

print(pd.__version__)
CSV_DATA = 'datasets/data.csv'
JSON_DATA = 'datasets/data.json'

2.1.4


In [26]:
my_dataset = {
  'cars': ["BMW", "Volvo", "Ford"],
  'passings': [3, 7, 2]
}
df = pd.DataFrame(my_dataset)
print(df)

    cars  passings
0    BMW         3
1  Volvo         7
2   Ford         2


## Series

* A Pandas Series is like a column in a table.
* It is a one-dimensional array holding data of any type.

In [27]:
arr = [1, 7, 2]
my_var = pd.Series(arr)
print(my_var)

0    1
1    7
2    2
dtype: int64


### Labels

* If nothing else is specified, the values are labeled with their index number. First value has index 0, second value has index 1 etc.
* This label can be used to access a specified value.

In [28]:
arr = [1, 7, 2]
my_var = pd.Series(arr)
print(my_var[0])

1


#### Create Labels

* With the index argument, you can name your own labels.
* When you have created labels, you can access an item by referring to the label.

In [29]:
a = [1, 7, 2]
my_var = pd.Series(a, index = ["x", "y", "z"])
print(my_var)
print(my_var["y"])

x    1
y    7
z    2
dtype: int64
7


### Key/Value Objects as Series

> The keys of the dictionary become the labels.

In [30]:
calories = {"day1": 420, "day2": 380, "day3": 390}
my_var = pd.Series(calories)
print(my_var)

day1    420
day2    380
day3    390
dtype: int64


In [31]:
# To select only some of the items in the dictionary, use the index argument and specify only the items you want to include in the Series.
calories = {"day1": 420, "day2": 380, "day3": 390}
my_var = pd.Series(calories, index = ["day1", "day2"])
print(my_var)

day1    420
day2    380
dtype: int64


## DataFrames

* A Pandas DataFrame is a 2-dimensional data structure, like a 2-dimensional array, or a table with rows and columns.
* Data sets in Pandas are usually multidimensional tables, called DataFrames.
* Series is like a column, a DataFrame is the whole table.


In [32]:
data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

df = pd.DataFrame(data)
print(df)

   calories  duration
0       420        50
1       380        40
2       390        45


### Locate Row

Pandas use the `loc` attribute to return one or more specified row(s)

In [33]:
data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

df = pd.DataFrame(data)
print(df.loc[0])  # Refer to the row index.
print(df.loc[[0, 1]])  # Use a list of indices.

calories    420
duration     50
Name: 0, dtype: int64
   calories  duration
0       420        50
1       380        40


### Named Indexes

* With the `index` argument, you can name your own indexes.

In [34]:
data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

df = pd.DataFrame(data, index = ["day1", "day2", "day3"])
print(df)

      calories  duration
day1       420        50
day2       380        40
day3       390        45


#### Locate Named Indexes

* If you have a large DataFrame with many rows, Pandas will only return the first 5 rows, and the last 5 rows:
* use `to_string()` to print the entire DataFrame.

In [35]:
print(df.loc["day2"])  # Refer to the named index.

calories    380
duration     40
Name: day2, dtype: int64


## Read CSV

In [46]:
df = pd.read_csv(CSV_DATA)
print(df)

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
..        ...    ...       ...       ...
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

[169 rows x 4 columns]


In [44]:
df = pd.read_csv(CSV_DATA)
print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

### max_rows

* The number of rows returned is defined in Pandas option settings.
* You can check your system's maximum rows with the `pd.options.display.max_rows` statement.
* In my system the number is `60`, which means that if the DataFrame contains more than `60` rows, the `print(df)` statement will return only the headers and the first and last 5 rows.
* You can change the maximum rows number with the same statement.

In [38]:
print(pd.options.display.max_rows)

9999


In [50]:
pd.options.display.max_rows = 9999
df = pd.read_csv(CSV_DATA)
print(df)
pd.options.display.max_rows = 60  # Converting back to 60.

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

# Read JSON

In [53]:
df = pd.read_json(JSON_DATA)
print(df)

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
..        ...    ...       ...       ...
164        60    105       140     290.8
165        60    110       145     300.4
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

[169 rows x 4 columns]


### Dictionary as JSON

* JSON = Python Dictionary
* JSON objects have the same format as Python dictionaries.
* If your JSON code is not in a file, but in a Python Dictionary, you can load it into a DataFrame directly.

In [54]:
data = {
  "Duration":{ "0":60, "1":60, "2":60, "3":45, "4":45, "5":60},
  "Pulse":{"0":110, "1":117, "2":103, "3":109, "4":117, "5":102},
  "Maxpulse":{"0":130, "1":145, "2":135, "3":175, "4":148, "5":127},
  "Calories":{"0":409, "1":479, "2":340, "3":282, "4":406, "5":300}
}

df = pd.DataFrame(data)
print(df)

   Duration  Pulse  Maxpulse  Calories
0        60    110       130       409
1        60    117       145       479
2        60    103       135       340
3        45    109       175       282
4        45    117       148       406
5        60    102       127       300


## Analyzing DataFrames

### Viewing the Data

* One of the most used method for getting a quick overview of the DataFrame, is the `head()` method.
* The `head()` method returns the headers and a specified number of rows, starting from the top.
* If the number of rows is not specified, the `head()` method will return the top 5 rows.
* There is also a `tail()` method for viewing the last rows of the DataFrame.
* The `tail()` method returns the headers and a specified number of rows, starting from the bottom.

In [56]:
df = pd.read_csv(CSV_DATA)
print(df.head(10))
print(df.tail())

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0
5        60    102       127     300.0
6        60    110       136     374.0
7        45    104       134     253.3
8        30    109       133     195.1
9        60     98       124     269.0
     Duration  Pulse  Maxpulse  Calories
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4


### Info About the Data

* The DataFrames object has a method called `info()`, that gives you more information about the data set.

In [57]:
df = pd.read_csv(CSV_DATA)
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    int64  
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  164 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 5.4 KB
None


#### Result Explained

* The result tells us there are 169 rows and 4 columns.
* And the name of each column, with the data type.
* The `info()` method also tells us how many Non-Null values there are present in each column, and in our data set it seems like there are 164 of 169 Non-Null values in the "Calories" column. Which means that there are 5 rows with no value at all, in the "Calories" column, for whatever reason.